jupyterhub/scopes.py

```diff
@@ -629,3 +630,95 @@ def describe_raw_scopes(raw_scopes, username=None):
             }
         )
     return descriptions
+
+
+_custom_scope_pattern = re.compile(r"^custom:[a-z0-9]+[a-z0-9_\-\*:]+[a-z0-9_\*]$")
+
+
+def define_custom_scopes(scopes):
+    """Define custom scopes
+
+    Scopes must start with `custom:`.
+    It is recommended to name custom scopes with a pattern like::
+
+        custom:$your-project:$action:$resource
+
+    e.g.::
+
+        custom:jupyter_server:read:contents
+
+    That makes them easy to parse and avoids collisions across projects.
+
+    All scopes must have at least a `definition`,
+    which will be displayed on the oauth authorization page,
+    and _may_ have a `subscopes` list of other scopes if having one scope
+    should imply having other, more specific scopes.
+
+    Args:
+
+    scopes: dict
+        A dictionary of scope definitions.
+        The keys are the scopes,
+        while the values are dictionaries with.
+        Custom scopes must start with `custom:`
+        and contain only lowercase ascii letters, numbers, hyphen, underscore, colon, and asterisk (-_:*).
+        The part after `custom:` must start with a letter or number.
+        Scopes may not end with a hyphen or colon.
+
+    Examples::
+
+        define_custom_scopes(
+            {
+                "custom:jupyter_server:read:contents": {
+                    "description": "read-only access to files in a Jupyter server",
+                },
+                "custom:jupyter_server:read": {
+                    "description": "read-only access to a Jupyter server",
+                    "subscopes": [
+                        "custom:jupyter_server:read:contents",
+                        "custom:jupyter_server:read:kernels",
+                        "...",
+                },
+            }
+        )
+    """
+    for scope, scope_definition in scopes.items():
+        if scope in scope_definitions and scope_definitions[scope] != scope_definition:
+            raise ValueError(
+                f"Cannot redefine scope {scope}={scope_definition}. Already have {scope}={scope_definitions[scope]}"
+            )
+        if not _custom_scope_pattern.match(scope):
+            raise ValueError(
+                f"Invalid scope: {scope!r}. Scopes must start with 'custom:' and contain only lowercase ascii letters, numbers, hyphen, underscore, colon, and asterisk."
```

Comment:
This is not a complete description and this is a text that should be updated in sync with changes to the pattern and the docstring that includes the description.

This is the text from the docstring that was missing:

```
(-_:*).
        The part after `custom:` must start with a letter or number.
        Scopes may not end with a hyphen or colon.
```

For anything that is hardcoded and repeated it would be good to have mentions on where we as maintainers need to keep things in sync - maybe we need one note here as well.

Reply:
Yeah, that's why I think it should not be a complete description here. I'll have another look at making things clear and concise and avoid duplications, and place comments where duplications cannot be avoided.